### PR TITLE
Prompt tool

### PR DESCRIPTION
**Pull request by albertodepaola**
# What does this PR do?

Adds a simple script that shows a Gradio interface to correctly format Llama 3 prompts. Additional models and version are upcoming.

Adds a simple notebook to be able to run on Colab as well.

## Feature/Issue validation/testing

Tested running the script and notebook and get a formatted prompt and code as well. Run the generated code to get the prompt again as well

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/llama-recipes/blob/main/CONTRIBUTING.md#pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?  
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: tools/prompt_generator/prompt.py [hunk 1]
```diff
@@ -56,6 +111,9 @@ def format_prompt_template_listener(system_prompt, user_prompt_1, assistant_resp
                         hf_output = gr.Textbox(label="Using HF Transformers", interactive=False, min_width=600, lines=25)
 
     system_prompt.change(format_prompt_template_listener, [system_prompt, user_prompt_1, assistant_response, user_prompt_2], [prompt_output, python_output, hf_output])
```
Comment (varunfb) [inline tools/prompt_generator/prompt.py]: Convert second and third parameter to a variable.

Comment (varunfb) [reply tools/prompt_generator/prompt.py]: Do the same thing for the remaining lines.

File: tools/prompt_generator/prompt.py [hunk 2]
```diff
@@ -90,17 +48,20 @@ def encode_dialog_prompt(dialog: Dialog) -> str:
             prompt_template = gr.Dropdown(["Single Turn", "Multi Turn"], label="Prompt Template", filterable=False)
 
             gr.Markdown("## Input Prompts")
-            system_prompt = gr.Textbox(label="System prompt", lines=2)
-            user_prompt_1 = gr.Textbox(label="User prompt", lines=2)
-            assistant_response = gr.Textbox(label="Assistant response", lines=2, visible=False, elem_id="assistant_response")
-            user_prompt_2 = gr.Textbox(label="User prompt", lines=2, visible=False, elem_id="user_prompt_2")
+            system_prompt = gr.Textbox(label="System prompt", lines=2, placeholder="Optional System prompt for the model")
+            user_prompt_1 = gr.Textbox(elem_id="user_prompt_1", label="User prompt *", lines=2, placeholder="Mandatory user prompt")
+            assistant_response = gr.Textbox(label="Assistant response", lines=2, visible=False, elem_id="assistant_response", placeholder="Mandatory assistant if prompt template is Multi Turn")
```
Comment (varunfb) [inline tools/prompt_generator/prompt.py]: Assistant prompt required on Multi Turn

File: tools/prompt_generator/prompt.py [hunk 3]
```diff
@@ -90,17 +48,20 @@ def encode_dialog_prompt(dialog: Dialog) -> str:
             prompt_template = gr.Dropdown(["Single Turn", "Multi Turn"], label="Prompt Template", filterable=False)
 
             gr.Markdown("## Input Prompts")
-            system_prompt = gr.Textbox(label="System prompt", lines=2)
-            user_prompt_1 = gr.Textbox(label="User prompt", lines=2)
-            assistant_response = gr.Textbox(label="Assistant response", lines=2, visible=False, elem_id="assistant_response")
-            user_prompt_2 = gr.Textbox(label="User prompt", lines=2, visible=False, elem_id="user_prompt_2")
+            system_prompt = gr.Textbox(label="System prompt", lines=2, placeholder="Optional System prompt for the model")
+            user_prompt_1 = gr.Textbox(elem_id="user_prompt_1", label="User prompt *", lines=2, placeholder="Mandatory user prompt")
```
Comment (varunfb) [inline tools/prompt_generator/prompt.py]: User prompt required.

File: tools/prompt_generator/prompt.py [hunk 4]
```diff
@@ -90,17 +48,20 @@ def encode_dialog_prompt(dialog: Dialog) -> str:
...
+            assistant_response = gr.Textbox(label="Assistant response", lines=2, visible=False, elem_id="assistant_response", placeholder="Mandatory assistant if prompt template is Multi Turn")
+            user_prompt_2 = gr.Textbox(label="User prompt *", lines=2, visible=False, elem_id="user_prompt_2", placeholder="Mandatory user prompt if prompt template is Multi Turn")
```
Comment (varunfb) [inline tools/prompt_generator/prompt.py]: User prompt required on Multi turn

File: tools/prompt_generator/prompt.py [hunk 5]
```diff
@@ -110,11 +71,6 @@ def encode_dialog_prompt(dialog: Dialog) -> str:
                     with gr.Tab("Hugging Face"):
                         hf_output = gr.Textbox(label="Using HF Transformers", interactive=False, min_width=600, lines=25)
 
-    system_prompt.change(format_prompt_template_listener, [system_prompt, user_prompt_1, assistant_response, user_prompt_2], [prompt_output, python_output, hf_output])
-    user_prompt_1.change(format_prompt_template_listener, [system_prompt, user_prompt_1, assistant_response, user_prompt_2], [prompt_output, python_output, hf_output])
-    assistant_response.change(format_prompt_template_listener, [system_prompt, user_prompt_1, assistant_response, user_prompt_2], [prompt_output, python_output, hf_output])
-    user_prompt_2.change(format_prompt_template_listener, [system_prompt, user_prompt_1, assistant_response, user_prompt_2], [prompt_output, python_output, hf_output])
-
-
+    submit.click(format_prompt_template_listener, [system_prompt, user_prompt_1, assistant_response, user_prompt_2], [notice, prompt_output, python_output, hf_output])
```
Comment (varunfb) [inline tools/prompt_generator/prompt.py]: Replace with variables.

File: tools/prompt_generator/prompt.py (hunk 1 above)
Comment (albertodepaola) [reply tools/prompt_generator/prompt.py]: I'm not sure I follow what do you mean with converting the second and thid param to a variable? Is is to put in the explicit `input` and `output` variable names in the call to the listener?

**Review by varunfb — CHANGES_REQUESTED** [review]
This is awesome. Ship it !!

**Review by varunfb — APPROVED** [review]
This is awesome. Ship it !!

**Pull request merged**